neet/automata/eca.py

```diff
@@ -654,19 +654,19 @@ def neighbors(self,size,index=None,direction='both'):
             ValueError: index must be <size if self.boundary==None
         """
         if direction == 'in':
-            if index:
+            if index is not None:
```

Comment:
thanks for catching these

Reply:
I've also been trying to think about the "all nodes having self-links" thing. I feel that there must be a way to write down conditions for a nodes future state which for most cases do not depend on the node itself (probably @dglmoore and @siyuzhou have thought about this more). I started to do this but maybe it's not a worthwhile endeavor. Perhaps @siyuzhou remove-useless-edges function gives him some insight here? This certainly seems like a problem related to finding alternate network topologies that yield the same dynamics.

Reply:
"All nodes having self-links" applies to WTNetwork because of the condition that when the weighted sum on a node is equal to its threshold, the next state remains the same as the previous one. This means the node still has to look at the state of itself occasionally, ie its state has potential effect on its future state. Earlier, I made it mandatory to include a "self dependency" when converting a WTNetwork to a LogicNetwork. Then I found sometimes the weighted sum is not possible to equal the threshold, so I added the step removing irrelevant links to clear-up the converted network. For a generic LogicNetwork, it is common to not have self-links for many nodes.

Reply:
Thanks, Siyu!  So now the question is: Do we have to run through the irrelevant links calculation every time we want to calculate neighbors for a weight threshold network?  If this is an involved calculation, we may want to do it once and save the results, but that starts to get a little involved...

Reply:
It might be worth assuming that if a weight is non-zero that the link is real. We could implement a reduction step for WTNetworks that removes non-functional links. Then this wouldn't be an assumption.

Reply:
As in: have an option to the neighbors function that specifies whether you want “quick and dirty” or “reduced”?  I think that could work...

Reply:
(Note that we don’t currently store information about whether links are functional, so if we implemented it this way the network itself wouldn’t be reduced, but we could ask for a reduced representation of the network’s neighbors.)